Building wxGTK/Win32 using MSys2 with configure fixes.

stahta01 opened the pull request:
First attempted fix:  "Fix to build wxGTK under Win32"   #415
Second attempted fix: "Building wxGTK/Win32 using MSys2" #426

I redid the code changes in this third attempt at an wxGTK/Win32 fix.
And, I added more configure fixes to stop the message when running the 
minimal sample. I included "docs/msw/msys2-gtk.txt" as the readme, vadz 
requested to added on the first attempted fix. And, I tested the 
minimal sample and it now works without issue.

Tim S.

Commit:
Only change wxUSE_CHECKLISTBOX to 0 when MSW toolkit.

Commit:
Configure changes needed for wxGTK/win32.

In configure moved WX_ARG_FEATURE(regkey outside of if.
Have configure under windows skip linking to X11.
Have configure check for "USE_WIN32" instead of for "wxUSE_MSW".
  wxUSE_FSWATCHER
  wxUSE_SNGLINST_CHECKER
  wxUSE_SOCKETS
  wxUSE_DIB

Commit:
Added directions to build wxGTK/Win32 using MSys2.

Commit:
When not Unix skip wxOVERRIDE in GetIconFromMimeType.

vadz commented on configure.in:
```diff
@@ -5947,10 +5950,8 @@ if test "$wxUSE_SOCKETS" = "yes"; then
-    dnl right now I'm assuming it also uses the winsock stuff
-    dnl like mingw does..  -- RL
-    if test "$TOOLKIT" != "MSW"; then
+    dnl under Windows we always have sockets
```
This comment doesn't seem to be appropriate here, it should probably say "We don't use getsockname()" under Windows or something like that.

stahta01 replied:
I will try to find a better comment for here.

Tim S.

vadz commented on configure.in:
```diff
@@ -7263,6 +7260,10 @@ if test "$USE_WIN32" = 1 ; then
     fi
 
+    if test "$wxUSE_DIB" = "yes"; then
+        AC_DEFINE(wxUSE_WXDIB)
```
Isn't `wxDIB` really wxMSW-specific? I.e. are we really going to use it in wxGTK code now? AFAICS so far we always use it in wxMSW code only and we have checks like `#if !defined(__WXMSW__) || wxUSE_WXDIB` in `src/html/m_image.cpp`.

stahta01 replied:
So, you think we should NOT use wxDIB under wxGTK/Win32. From the code ww will have to NOT define wxHAS_IMAGES_IN_RESOURCES or change wxHAS_IMAGES_IN_RESOURCES so wxUSE_WXDIB is NOT needed to be defined to use images in resources.

I just wish to confirm you do NOT think we should use both wxDIB and "images in resources" in wxGTK/Win32?

Tim S.

vadz replied:
I'm not actually completely sure what's the philosophy of wxGTK/Win32 port, but I think that the goal is to be closer to GTK+ than to MSW and from this point of view, neither `wxDIB` nor images in the resources should be used in it.

@kosenko Please correct me if I'm wrong.

stahta01 replied:
I will work on trying to get the minimal sample to run without issue.
With wxUSE_WXDIB=0 and with wxHAS_IMAGES_IN_RESOURCES undefined.

@kosenko Please tell me if this is NOT the correct choice.

Tim S.

stahta01 replied:
In file "wx/gtk/setup0.h" wxUSE_WXDIB is set to 1.
So, I am stopping work on this change till kosenko responds.

Tim S.

stahta01 replied:
I think I have figured out the correct solution.
I plan to add wxUSE_IMAGES_IN_RESOURCES to select whether wxICON looks for xpm or resource file under windows. The main code changes will be in "wx/gdicmn.h"; but, I will have to change most places that check wxHAS_IMAGES_IN_RESOURCES to instead check wxUSE_IMAGES_IN_RESOURCES. So, nearly every sample will need changed and a few demos and utils. Tim S.

stahta01 replied:
I have decided to add this feature after finishing this pull commits.
From http://trac.wxwidgets.org/ticket/13902#comment:32
kosenko wrote "Enabled some Windows features: DIB ..."

The work is big enough for a separate pull.

Tim S.

stahta01 replied:
I thought up how to change to using wxUSE_IMAGES_IN_RESOURCES without as many code changes as I thought. I will have wxUSE_IMAGES_IN_RESOURCES=1 under WindowsOS else it will be zero. And, I will define wxHAS_IMAGES_IN_RESOURCES when it is 1; therefore will NOT have to update all the samples.

@vadz Is this adding of wxUSE_IMAGES_IN_RESOURCES approved by you?

I plan to have setup.h define it to 1 when it is undefined to allow the client to set whether it is 1 or 0.

Tim S.

stahta01 replied:
@vadz

Can you commit the changes I have done.
I do NOT have the time right now to figure out how to do the images and relearn how to edit the doc files and setup.h files.

Thank you.

Tim S.

vadz commented on include/wx/gtk/mimetype.h:
```diff
@@ -23,7 +23,11 @@
+#if defined(__UNIX__)
     wxString GetIconFromMimeType(const wxString& mime) wxOVERRIDE;
+#else
+    wxString GetIconFromMimeType(const wxString& mime);
```
Why do we need it in non-Unix case? Is it going to be used at all? Under Unix it's used as a hook for the GUI toolkit to provide access to GUI-specific functionality, but this doesn't seem to make any sense under MSW...

stahta01 replied:
I will have to do code testing to confirm if it is needed or not.

Tim S.

stahta01 replied:
It fails to work; to fix it I would have to return to code that was reject the first time by kosenko.

@kosenko What do you think about this change request?

Tim S.

vadz replied:
Sorry, why does it fail to work? AFAICS the only code using this function (`GetIconFromMimeType()`) is in Unix code, so either it should be declared in the base class (and then `override` can be used) or it shouldn't be used at all.

So what exactly goes wrong if you omit the `#else` branch entirely?

stahta01 replied:
It is used in wxGTKMimeTypesManagerImpl inside src/gtk/mimetype.cpp.

Tim S.

stahta01 replied:
../src/gtk/mimetype.cpp:22:77: error: no 'wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString&)' member function declared in class 'wxGTKMimeTypesManagerImpl'
 wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString& mime)
                                                                             ^
../src/gtk/mimetype.cpp:54:55: error: expected declaration before end of line
         wxGCC_WARNING_SUPPRESS(deprecated-declarations)

stahta01 replied:
I have ran sample typetest and it works without any code changes. No idea how to test GetIconFromMimeType under wxGTK/Win32.
Without a way to test I am unsure what to do about this change request.

Tim S.

stahta01 replied:
I decided that if sample typetest works after the changes I will do your requested change.

Tim S.

stahta01 replied:
@vadz I have committed the changes to only use  GetIconFromMimeType under Unix.

Tim S.

vadz commented on include/wx/msw/chkconf.h:
```diff
@@ -183,7 +183,7 @@
 #endif
 
-#if !wxUSE_OWNER_DRAWN && !defined(__WXUNIVERSAL__)
+#if !wxUSE_OWNER_DRAWN && !defined(__WXUNIVERSAL__) && defined(__WXMSW__)
```
I think this file implicitly assumes being used only for wxMSW, so it would be nice to make it more clear that it can now be used with the other ports too, e.g. like this:

    /* wxMSW-specific checks: notice that this file is also used with wxUniv and
       can even be used with wxGTK, when building it under Windows. */
    #if defined(__WXMSW__) && !defined(__WXUNIVERSAL__)
         ....
    #endif

stahta01 replied:
I will rewrite this check in the style suggested.

Tim S.

vadz reviewed (commented):
Thanks, it will be nice to have this working with configure and the new README is definitely useful!

However I do have some doubts about a couple of changes here which it would be nice to clear before merging.

Commit:
Edited when setting wxUSE_CHECKLISTBOX to 0 under MSW port.

Commit:
Edited dnl comment in "configure.in".

kosenko commented:
@stahta01 it looks like you have a problem because you are trying to build wxGTK/Win32 using MinGW with wxWidgets/configure file instead of wxWidgets/master/build/msw/makefile.gcc file.

wxGTK/Win32 port is designed to be like wxMSW port, but use GTK+/Win32 instead of Windows Controls. So wxGTK/Win32 is Windows-platform based, not Cygwin-like (Unix-under-Windows) platform based.
Visual C++ (and raw MingGW) doesn't provide Unix stuff by default, so it works fine.
Philosophy of wxGTK/Win32 port is to be closer to GTK+ and Windows platform, but not Cygwin-platform.
So it uses wxDIB, OLE, display (temporarily, while native GTK+ support is weak), dialup, app & browser launch (to open documents and urls), sound, joystick from Windows platform.

wxDIB is required by wxGTK/Win32 to load images from rc resource files that are used in Visual C++ project files. wxDIB is not required in wxGTK/X11/Cygwin build.
wxGTK/Win32 uses Windows-platform mimetypes support because Unix aproach doesn't work with Visual C++.

Using `configure` to build wxGTK supposes Unix libs and wxGTK/Win32 not supports this way. Hypothetical it is possible to implement and wxGTK/Win32 could use Cygwin-platform instead of Windows-platform. But it isn't easy.
Do you have build minimal sample? It runs? Other samples runs? Do you have screenshots like http://trac.wxwidgets.org/ticket/13902#comment:32 ?

stahta01 commented:
I will work on adding some images from a few samples.
I am NOT good with images; so, it might take a while to add images to this pull.
I am waiting for feedback before doing much more than that on this pull request.

In future pulls:

I will work on adding wxUSE_IMAGES_IN_RESOURCES to improve ease of wxGTK code ports to wxGTK/Win32; but, I decided it should default to enabled under wxGTK/Win32.

I will also work on "keyboard handling" because it is needed before I can work on wxUIActionSimulator under wxGTK/Win32.

I will work on fixing wxMSW/MSys2 non GUI test unit code before starting work on "keyboard handling" and wxUIActionSimulator under wxGTK/Win32. I have no idea how to fix the MinGW64 CRT lib bug that is likely causing the issue under MSys2.

Tim S.

Commit:
Added samples to test to wxGTK MSys2 directions.

Also, added option "--disable-precomp-headers".
And, changed MSys2 packages to install.

Commit:
Added MSys2 wxMSW configure build directions.

Commit:
Do NOT have method GetIconFromMimeType except under Unix.

kosenko commented on configure.in:
```diff
@@ -749,9 +749,10 @@ WX_ARG_FEATURE(threads,     [  --enable-threads        use threads], wxUSE_THREA
-WX_ARG_FEATURE(regkey,      [  --enable-regkey         use wxRegKey class (Win32 only)], wxUSE_REGKEY)
 fi
 
+WX_ARG_FEATURE(regkey,      [  --enable-regkey         use wxRegKey class (Win32 only)], wxUSE_REGKEY)
```
wxRegKey isn't available on all platforms. It (and probably wxIniConfig, and dbghelp) was wxMSW-port dependent and probably should become Windows-platform dependent.

stahta01 replied:
Problem is at this point in the configure it does NOT know it is under Windows or Unix.
Maybe a bakefile/configure expert can do it; but, I wasted several hours trying to figure it out.

Tim S.

kosenko replied:
It looks like either regkey check should be placed after "Platform specific tests" section (`USE_WIN32` definition) or this section should be moved before compile options. @vadz what do you think?

kosenko commented:
It seems this patchset will not break wxGTK/Win32 build using Visual C++.

It would be nice to have also shared (DLL) build to ensure that all functions have implementations. But as a first try static build is good enough.

Also it would be nice to have several samples compilable and runable, at least minimal and widgets samples.

autoconf `configure` build probably will have less features implemented then Visual C++ build because wxGTK in Makefile.in has no corresponding files.

stahta01 commented:
The wxWidgets builds under MSys2 packages using DLL without problem.
I just can NOT figure how to build the samples as DLLs and have them run.

Tim S.

vadz commented:
@kosenko Just to be sure, when you write that "it seems this will not break <...>", does it mean that you see no problems with merging this? If so, I'll just do it, even though I still didn't have time to test it myself. Of course, if anybody else is interested in testing this (please see the provided instructions -- thanks @stahta01!), it would be great too, TIA!

kosenko commented:
@vadz, Yes, please apply this patch.

vadz commented:
Thanks again!
